src/digraph.rs

```diff
@@ -273,85 +273,203 @@ impl PyDiGraph {
     }
 
     fn __getstate__(&self, py: Python) -> PyResult<PyObject> {
-        let out_dict = PyDict::new(py);
-        let node_dict = PyDict::new(py);
-        let mut out_list: Vec<PyObject> = Vec::with_capacity(self.graph.edge_count());
-        out_dict.set_item("nodes", node_dict)?;
-        out_dict.set_item("nodes_removed", self.node_removed)?;
-        out_dict.set_item("multigraph", self.multigraph)?;
-        let dir = petgraph::Direction::Incoming;
-        for node_index in self.graph.node_indices() {
-            let node_data = self.graph.node_weight(node_index).unwrap();
-            node_dict.set_item(node_index.index(), node_data)?;
-            for edge in self.graph.edges_directed(node_index, dir) {
-                let edge_w = edge.weight();
-                let triplet = (edge.source().index(), edge.target().index(), edge_w).to_object(py);
-                out_list.push(triplet);
-            }
+        let mut nodes: Vec<PyObject> = Vec::with_capacity(self.graph.node_count());
+        let mut edges: Vec<PyObject> = Vec::with_capacity(self.graph.edge_bound());
+
+        // save nodes to a list along with its index
+        for node_idx in self.graph.node_indices() {
+            let node_data = self.graph.node_weight(node_idx).unwrap();
+            nodes.push((node_idx.index(), node_data).to_object(py));
+        }
+
+        // edges are saved with none (deleted edges) instead of their index to save space
+        for i in 0..self.graph.edge_bound() {
+            let idx = EdgeIndex::new(i);
+            let edge = match self.graph.edge_weight(idx) {
+                Some(edge_w) => {
+                    let endpoints = self.graph.edge_endpoints(idx).unwrap();
+                    (endpoints.0.index(), endpoints.1.index(), edge_w).to_object(py)
+                }
+                None => py.None(),
+            };
+            edges.push(edge);
         }
-        let py_out_list: PyObject = PyList::new(py, out_list).into();
-        out_dict.set_item("edges", py_out_list)?;
-        Ok(out_dict.into())
+
+        let outdict = PyDict::new(py);
+        let nodes_lst: PyObject = PyList::new(py, nodes).into();
+        let edges_lst: PyObject = PyList::new(py, edges).into();
+        outdict.set_item("nodes", nodes_lst)?;
+        outdict.set_item("edges", edges_lst)?;
+        outdict.set_item(
+            "node_removed",
+            (self.graph.node_bound() != self.graph.node_count()).to_object(py),
```

Comment:
I actually don't know exactly how `node_removed` is implemented, but I saw the [original implementation](https://github.com/Qiskit/retworkx/blob/add8e5b8ad9026551acd2ac7eda1bd955ca656c1/src/digraph.rs#L323) also reset `node_removed` so I assume that there is no guarantee that `node_removed` is always correct and needed recalculation (e.g., delete n node and add back n node, `node_removed` is not set to false).

Reply:
It is enforced at retworkx's API. So if things are working on petgraph objects directly the attribute isn't updated for us. But if you use the methods on the pyclass it would. So for `__setstate__` we had to set this manually because we create a petgraph graph directly. But from the python api perspective we should always be tracking this correctly (if not it's a bug).